### PR TITLE
Allow import assertions on esm imports under module: nodenext

### LINKED ISSUE #46372: Import assertions not allowed in nodenext
I tried using import assertions while using `"module": "nodenext"`.

```ts
// @module: nodenext

import json from "some-module" assert { type: "json" };
//                             ~~~~~~~~~~~~~~~~~~~~~~~
// error:
// Import assertions are only supported when the '--module' option is set to 'esnext'.
```

**Expected**: no error

**Actual**:

```
Import assertions are only supported when the '--module' option is set to 'esnext'.
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #46372

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -3353,6 +3353,10 @@
         "category": "Error",
         "code": 2835
     },
+    "Import assertions are not allowed on statements that transpile to commonjs require calls.": {
```
Comment (andrewbranch) [inline src/compiler/diagnosticMessages.json:3356]:
```suggestion
    "Import assertions are not allowed on statements that transpile to CommonJS 'require' calls.": {
```

File: tests/baselines/reference/nodeModulesImportAssertions(module=node12).errors.txt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+==== tests/cases/conformance/node/index.ts (2 errors) ====
+    import json from "./package.json" assert { type: "json" };
+                     ~~~~~~~~~~~~~~~~
+!!! error TS7062: JSON imports are experimental in ES module mode imports.
+                                      ~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS2821: Import assertions are only supported when the '--module' option is set to 'esnext'.
+==== tests/cases/conformance/node/otherc.cts (3 errors) ====
+    import json from "./package.json" assert { type: "json" }; // should error, cjs mode imports don't support assertions
+                                      ~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS2821: Import assertions are only supported when the '--module' option is set to 'esnext'.
```
Comment (andrewbranch) [inline tests/baselines/reference/nodeModulesImportAssertions(module=node12).errors.txt:17]: Looks like this error text needs to be updated to include `nodenext`

File: tests/baselines/reference/nodeModulesImportAssertions(module=node12).errors.txt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    const json2 = import("./package.json", { assert: { type: "json" } }); // should be fine
+                         ~~~~~~~~~~~~~~~~
+!!! error TS7062: JSON imports are experimental in ES module mode imports.
+                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS1324: Dynamic imports only support a second argument when the '--module' option is set to 'esnext'.
```
Comment (andrewbranch) [inline tests/baselines/reference/nodeModulesImportAssertions(module=node12).errors.txt:22]: And this one

**Review by andrewbranch — CHANGES_REQUESTED** [review]

**Review by gabritto — APPROVED** [review]
As someone who doesn't know that much about modules and import assertions, the error message that says "Import assertions are not allowed on statements that transpile to commonjs 'require' calls" looks a bit puzzling, even though it's accurate. But I don't know how that could be improved; maybe instead of saying "statements that transpile to CommonJs..." we could explicitly list the scenarios when this is true? That might be too verbose and not helpful either...

Otherwise this looks good.

**Review by andrewbranch — APPROVED** [review]

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick this to release-4.5

**Comment by weswigham** [comment]
> could explicitly list the scenarios when this is true?

I think "import declarations and external reexport declarations in a cjs mode file except for dynamic import expressions" is a bit of a mouthful. And probably not much better.

**Pull request merged**